### PR TITLE
MULE-8987 Improve truncated stack trace in exception log

### PR DESCRIPTION
**Pull request by elrodro83**

### CODE REVIEW

**Comment by pablokraan** [comment]
Fix code style and add unit tests

File: core/src/main/java/org/mule/config/ExceptionHelper.java [hunk 1]
```diff
@@ -541,34 +543,63 @@ public static Map getExceptionInfo(Throwable t)
...
+        Throwable root = getRootException(t);
+        MuleException rootMule = getRootMuleException(t);
+
         StringBuilder buf = new StringBuilder();
-        // get exception stack
-        List<Throwable> exceptions = getExceptionsAsList(t);
 
-        int i = 1;
-        for (Iterator<Throwable> iterator = exceptions.iterator(); iterator.hasNext(); i++)
+        ExceptionReader er = getExceptionReader(rootMule);
```
Comment (pablokraan) [inline core/src/main/java/org/mule/config/ExceptionHelper.java]: Rename "er" to something more meaningful

File: core/src/main/java/org/mule/config/ExceptionHelper.java [hunk 2]
```diff
@@ -69,7 +71,7 @@
     private static Map<String,Properties> errorMappings = new HashMap<String,Properties>();
     private static Map<String,Boolean> disposeListenerRegistered = new HashMap<String,Boolean>();
 
-    private static int exceptionThreshold = 0;
+    private static int exceptionThreshold = 3;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/config/ExceptionHelper.java]: This seems to be a constant. Make it final and rename to EXCEPTION_THRESHOLD

File: core/src/main/java/org/mule/config/ExceptionHelper.java [hunk 3]
```diff
@@ -541,34 +543,63 @@ public static Map getExceptionInfo(Throwable t)
...
         {
-            if (i > exceptionThreshold && exceptionThreshold > 0)
+            int processedElements = 0;
+            int nonMuleElements = 0;
+            int processedMuleElements = 1;
+            for (StackTraceElement stackTraceElement : root.getStackTrace())
             {
-                buf.append("(").append(exceptions.size() - i + 1).append(" more...)");
-                break;
+                if (processedMuleElements > exceptionThreshold && exceptionThreshold > 0)
```
Comment (pablokraan) [inline core/src/main/java/org/mule/config/ExceptionHelper.java]: exceptionThreshold > 0 is redundant here. If you want to be sure of that, then check it on the class initialization

File: core/src/main/java/org/mule/config/ExceptionHelper.java [hunk 4]
```diff
@@ -541,34 +543,63 @@ public static Map getExceptionInfo(Throwable t)
...
+                            .append(" more...)")
+                            .append(SystemUtils.LINE_SEPARATOR);
+                    }
+                    
+                    buf.append("  ")
+                        .append(stackTraceElement.getClassName())
+                        .append(":")
+                        .append(stackTraceElement.getLineNumber())
+                        .append(SystemUtils.LINE_SEPARATOR);
+                    nonMuleElements = 0;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/config/ExceptionHelper.java]: Why do you need to reset the number of nonMuleElements?

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java [hunk 5]
```diff
@@ -80,9 +85,70 @@ public void testGetNonMuleExceptionCause()
         assertThat(ExceptionHelper.getNonMuleException(new IllegalArgumentException()),IsInstanceOf.instanceOf(IllegalArgumentException.class));
     }
 
+    @Test
+    public void testFilteredStack1()
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: Don't use testFilteredStack1, testFilteredStack2, etc. Use test names that describe the difference between them

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java [hunk 6]
```diff
@@ -80,9 +85,70 @@ public void testGetNonMuleExceptionCause()
...
+        }
+        catch (Exception e)
+        {
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("foo (org.mule.api.DefaultMuleException)"));
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("  org.mule.test.config.ExceptionHelperTestCase:"));
+            assertThat(ExceptionHelper.getExceptionStack(e), not(containsString("  (0 more...)")));
+        }
+    }
+
+    private void recursive(int calls, Closure closure)
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: Rename to make it more clear, like "generateStackTrace(int recursiveLevels...)

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java [hunk 7]
```diff
@@ -80,9 +85,70 @@ public void testGetNonMuleExceptionCause()
...
+                        }
+                    });
+                }
+            });
+        }
+        catch (Exception e)
+        {
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("foo (org.mule.api.DefaultMuleException)"));
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("  org.mule.test.config.ExceptionHelperTestCase$1$1:"));
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("  (1 more...)")); // CollectionUtils.forAllDo
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: I see now why you need to reset the nonMuleelementsCounter

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java [hunk 8]
```diff
@@ -80,9 +85,70 @@ public void testGetNonMuleExceptionCause()
...
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("  org.mule.test.config.ExceptionHelperTestCase$1:"));
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("  org.mule.test.config.ExceptionHelperTestCase:"));
+            assertThat(ExceptionHelper.getExceptionStack(e), containsString("  (" + (calls + 13) + " more...)")); // recursive
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: Can you create a matcher that receives a the expected strings and compares that against the exception's stack trace?

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java [hunk 9]
```diff
@@ -80,9 +85,70 @@ public void testGetNonMuleExceptionCause()
...
+                    CollectionUtils.forAllDo(Collections.singleton(null), new Closure()
+                    {
+                        @Override
+                        public void execute(Object input)
+                        {
+                            throw new RuntimeException(new DefaultMuleException(MessageFactory.createStaticMessage("foo")));
+                        }
+                    });
+                }
+            });
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: need a fail here

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java (hunk 7 above)
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: The stack trace must contain 3 lines of org.mule code, is the test data wrong?

File: core/src/main/java/org/mule/config/ExceptionHelper.java (hunk 4 above)
Comment (elrodro83) [inline core/src/main/java/org/mule/config/ExceptionHelper.java]: Count has to start again after passing through a mule code entry.

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java [hunk 10]
```diff
@@ -80,9 +87,116 @@ public void testGetNonMuleExceptionCause()
         assertThat(ExceptionHelper.getNonMuleException(new IllegalArgumentException()),IsInstanceOf.instanceOf(IllegalArgumentException.class));
     }
 
+    @Test
+    public void testFilteredStackIncludingNonMuleCode()
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: for consistency, remove "test" prefix

File: tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java [hunk 11]
```diff
@@ -80,9 +87,116 @@ public void testGetNonMuleExceptionCause()
...
+        {
+            closure.execute(null);
+        }
+        else
+        {
+            generateStackEntries(--calls, closure);
+        }
+    }
+
+    private static final class StringByLineMatcher extends BaseMatcher<String>
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/config/ExceptionHelperTestCase.java]: Extend from TypeSafeMatcher

**Pull request merged**